multisite/hacks.py

```diff
@@ -35,6 +36,18 @@ def SiteManager_clear_cache(self):
     models.SITE_CACHE.clear()
 
 
+# Override SiteManager._get_site_by_id
+def SiteManager_get_site_by_id(self, site_id):
+    """Patch _get_site_by_id to return the site from the DB if necessary."""
+    models = sys.modules.get(self.__class__.__module__)
+    if site_id not in models.SITE_CACHE:
+        site = self.get(pk=site_id)
+        models.SITE_CACHE[site_id] = site
+    # there can be a race condition between processes; if we can't find the site
+    # in the cache at this point, fetch it from the DB
+    return models.SITE_CACHE[site_id] or self.get(pk=site_id)
```

Comment:
`models.SITE_CACHE[site_id]` comes from `DictCache.__getitem__` which does now return None.  But, now I'm thinking maybe it would be better to put that back so it returns the KeyError as you'd expect from a dict, and get the site like this instead:
```
site = models.SITE_CACHE.get(site_id)
if site is None:
    site = self.get(pk=site_id)
    models.SITE_CACHE[site_id] = site
return site
```
Then we get the site from the cache at the beginning and return that; we don't check if the key is there at the beginning and then fetch it from the cache at the end, so we shouldn't run into the race condition.